src/main/java/com/iota/iri/TransactionValidator.java

```diff
@@ -119,25 +129,29 @@ public boolean checkSolidity(Hash hash, boolean milestone) throws Exception {
         while ((hashPointer = nonAnalyzedTransactions.poll()) != null) {
             if (analyzedHashes.add(hashPointer)) {
                 final TransactionViewModel transaction = TransactionViewModel.fromHash(tangle, hashPointer);
-                if(!transaction.isSolid()) {
-                    if (transaction.getType() == TransactionViewModel.PREFILLED_SLOT && !hashPointer.equals(Hash.NULL_HASH)) {
-                        transactionRequester.requestTransaction(hashPointer, milestone);
-                        solid = false;
-                        break;
-                    } else {
-                        if (solid) {
+                switch (transaction.subtangleStatus()) {
+                    case UNKNOWN:
+                        if (transaction.getType() == TransactionViewModel.PREFILLED_SLOT && !hashPointer.equals(Milestone.INITIAL_MILESTONE_HASH)) {
+                            transactionRequester.requestTransaction(hashPointer, priority);
+                            nonAnalyzedTransactions.clear();
+                            analyzedHashes.clear();
+                            return false;
+                        } else {
                             nonAnalyzedTransactions.offer(transaction.getTrunkTransactionHash());
                             nonAnalyzedTransactions.offer(transaction.getBranchTransactionHash());
                         }
-                    }
+                        break;
+                    case INVALID:
+                        if(!priority) {
```

Comment:
if this intended to allow `priority` to change `INVALID` to `SOLID`,
then it will only do so for the first "level" of invalids. as the traversal is only done for `UNKNOWN`.

this could also lead to false marking as solid:
think of `checkSolidity` called from `Milestone.updateLatestSolidSubtangleMilestone` (L227) where `priority` is `true`
if the starting transaction was marked as `INVALID`, then the while loop would end w/o traversing & mark the milestone as `SOLID` (`updateSolidTransactions`) w/o actually traversing & checking.